src/utmpx.rs

```diff
@@ -0,0 +1,234 @@
+//!
+//! # Safety
+//!
+//! Standard-defined `utmpx` functions are explicitly **NOT** thread-safe.
+//! The caller must make sure to operate only on one `Utmp` singleton per process.
+
+use crate::errno::ErrnoSentinel;
+use crate::sys::time::TimeVal;
+use crate::unistd::Pid;
+use crate::{Errno, Error, Result};
+use std::convert::TryFrom;
+use std::ffi::CStr;
+use std::marker::PhantomData;
+
+/// Valid `UtmpEntry` entry types.
+#[derive(Clone, Copy, PartialEq, Eq, Debug, Hash, Ord, PartialOrd)]
+#[repr(i16)]
+pub enum EntryType {
+    /// No valid user accounting information.
+    Empty = libc::EMPTY,
+    /// Change in system run-level.
+    #[cfg(target_os = "linux")]
+    RunLevel = libc::RUN_LVL,
+    /// Identifies time of system boot.
+    BootTime = libc::BOOT_TIME,
+    /// Identifies time of system shutdown.
+    #[cfg(any(
+        target_os = "dragonfly",
+        target_os = "freebsd",
+        target_os = "ios",
+        target_os = "macos"
+    ))]
+    ShutdownTime = libc::SHUTDOWN_TIME,
+    /// Identifies time after system clock changed.
+    NewTime = libc::NEW_TIME,
+    /// Identifies time when system clock changed.
+    OldTime = libc::OLD_TIME,
+    /// Identifies a process spawned by the init process.
+    InitProcess = libc::INIT_PROCESS,
+    /// Identifies the session leader of a logged-in user.
+    LoginProcess = libc::LOGIN_PROCESS,
+    /// Identifies a process.
+    UserProcess = libc::USER_PROCESS,
+    /// Identifies a session leader who has exited.
+    DeadProcess = libc::DEAD_PROCESS,
+    /// Accounting.
+    #[cfg(target_env = "gnu")]
+    Accounting = libc::ACCOUNTING,
+}
+
+impl TryFrom<i16> for EntryType {
+    type Error = Error;
+
+    /// Try to build an `EntryType` from its discriminant.
+    fn try_from(value: i16) -> Result<Self> {
+        match value {
+            libc::EMPTY => Ok(EntryType::Empty),
+            #[cfg(target_os = "linux")]
+            libc::RUN_LVL => Ok(EntryType::RunLevel),
+            libc::BOOT_TIME => Ok(EntryType::BootTime),
+            #[cfg(any(
+                target_os = "dragonfly",
+                target_os = "freebsd",
+                target_os = "ios",
+                target_os = "macos"
+            ))]
+            libc::SHUTDOWN_TIME => Ok(EntryType::ShutdownTime),
+            libc::OLD_TIME => Ok(EntryType::OldTime),
+            libc::NEW_TIME => Ok(EntryType::NewTime),
+            libc::USER_PROCESS => Ok(EntryType::UserProcess),
+            libc::INIT_PROCESS => Ok(EntryType::InitProcess),
+            libc::LOGIN_PROCESS => Ok(EntryType::LoginProcess),
+            libc::DEAD_PROCESS => Ok(EntryType::DeadProcess),
+            #[cfg(target_env = "gnu")]
+            libc::ACCOUNTING => Ok(EntryType::Accounting),
+            _ => Err(Error::invalid_argument()),
+        }
+    }
+}
+
+/// Utmp login record.
+#[derive(Clone, PartialEq, Eq, Debug, Hash)]
+pub struct UtmpEntry {
+    ut_host: String,
+    ut_id: Vec<u8>,
```

Comment:
Since it has a fixed (platform-dependent) size, you can store `ut_id` as an array instead of a vec.

Reply:
To the best of my knowledge, the size of this field is not a public API and there is no stable `libc` constant defining it.
But I fear I may be missing some bits of your suggestion. If you have clues in other directions, and actual working declaration you want me to use here, I'll be happy to follow.